auth: update x509 endpoint for browser clients #6048

maany opened the pull request:
Fix #6048 ( solves the auth server aspect of the issue )

The new webui requires the x509 endpoint to be updated to support browser clients. This commit updates the endpoint to support both browser clients and CLI clients.

The /auth/x509 endpoint can return HTTP 206 Partial Content to indicate that multiple accounts were found for the given x509 identity and can work with the new WebUI to pick an account.

The /auth/x509 endpoint can return more detailed error messages to the webui (identity does not exist, identity not mapped to an account, ..).

maany committed:
auth: update x509 endpoint for browser clients #6048

The new webui requires the x509 endpoint to be updated to support
browser clients. This commit updates the endpoint to support both
browser clients and CLI clients.

The /auth/x509 endpoint can return HTTP 206 Partial Content to indicate
that multiple accounts were found for the given certificate and can
work with the new WebUI to pick an account.

The /auth/x509 endpoint can return more detailed error messages to the
webui ( idenity does not exist, identity not mapped to an account, ..).

bari12 commented on lib/rucio/web/rest/flaskapi/v1/auth.py:
```diff
@@ -943,14 +943,14 @@ def get_headers(self) -> "Optional[HeadersType]":
         ---
-        summary: x509 Allow cross-site scripting
-        description: x509 Allow cross-site scripting. Explicit for Authentication.
+        summary: x509_web Allow cross-site scripting
```
Isn't this just staying `x509`?

bari12 commented on lib/rucio/web/rest/flaskapi/v1/auth.py:
```diff
@@ -943,14 +943,14 @@ def get_headers(self) -> "Optional[HeadersType]":
-        summary: x509 Allow cross-site scripting
-        description: x509 Allow cross-site scripting. Explicit for Authentication.
+        summary: x509_web Allow cross-site scripting
+        description: x509_web Allow cross-site scripting. Explicit for Authentication.
```
Same

bari12 commented on lib/rucio/web/rest/flaskapi/v1/auth.py:
```diff
@@ -984,8 +984,8 @@ def options(self):
         ---
-        summary: x509
-        description: Authenticate a Rucio account via x509.
+        summary: x509_web
```
Also just `x509`?

bari12 commented on lib/rucio/web/rest/flaskapi/v1/auth.py:
```diff
@@ -984,8 +984,8 @@ def options(self):
-        summary: x509
-        description: Authenticate a Rucio account via x509.
+        summary: x509_web
+        description: Authenticate a Rucio account via x509. Works only for web browsers.
```
Scratch?

bari12 reviewed (commented):
Please update the initial ticket description, this is not current anymore (Since we do not create a new endpoint)

I think this is correct, I wonder though if the workflow should be different. We do more database queries this way since you always do `verify_identity` first, and then later on a `get_auth_token_x509`. I wonder if this should be done the other way around. Try `get_auth_token_x509` first, which will fail if no account is set and no default account is defined. and then go for the block you added basically.

mlassnig commented on lib/rucio/tests/common.py:
```diff
@@ -18,8 +18,9 @@
 import tempfile
-from random import choice
-from string import ascii_uppercase
+from typing import Optional
```
t follows after r, s in the alphabet :-p

mlassnig commented on lib/rucio/tests/common.py:
```diff
@@ -38,23 +39,27 @@
     """Return True if influxdb is available, else return False."""
     try:
         response = requests.get('http://localhost:8086/ping')
         if response.status_code == 204:
```
all these can probably be rewritten shorter like:

`return response.status_code == 204`

maany commented:
@bari12 Converting to a draft PR and I will discuss this tomorrow in person during our meeting. I have rewritten big chunks of the [webui](https://github.com/maany/rucio-webui-nextjs), which enable a more secure x509 workflow for browser clients. Essentially the rucio-token is never stored or made available to the browser unencrypted and is only available in an encrypted cookie. If we go this route, which is more secure, it might require reworking this PR a bit.

maany committed:
auth: export X-Rucio-Token-Expires header in x509 response #6048

maany committed:
code-quality: sort imports and reduce lines of code #6048

The rucio.tests.common file is being addressed in #6048 to reduce
type annotation errors. The commit addresses the comments from
@mlassnig in #6053 about sorting import block alphabetically
and simplifying the return statements.

maany committed:
web: enable GET on DIDs endpoint in blueprint
